### ISSUE TITLE
🐛 [iPadOS] If app is set to landscape only, vision camera gets preview orientation wrong

### ISSUE THREAD

**Issue by zzz08900** [cuvent/react-native-vision-camera#1455]
### What were you trying to do?

We have an iPadOS-only app that only supports landscape orientation. We tried to use react-native-vision-camera@2.15.4 with this app but found the preview is displayed with wrong orientation - it was rotated 90 degrees.
So when the iPad moves left and right, the image in react-native-vision-camera preview moves up and down.

### What happened instead?

With an app that only supports landscape orientation, when the iPad moves left and right, the image in react-native-vision-camera preview moves up and down.

### Device

iPad (9th Generation)

### VisionCamera Version

2.15.4

### Additional information

- [ ] I am using Expo
- [X] I have read the [Troubleshooting Guide](https://mrousavy.github.io/react-native-vision-camera/docs/guides/troubleshooting)
- [X] I agree to follow this project's [Code of Conduct](https://github.com/mrousavy/react-native-vision-camera/blob/main/CODE_OF_CONDUCT.md)
- [X] I searched for [similar issues in this repository](https://github.com/mrousavy/react-native-vision-camera/issues) and found none.

**Comment by zzz08900** [comment]
This bug is not present in version 2.14.x, I highly suspect this piece of code is what causing this issue.
![Screenshot from 2023-02-06 14-04-02](https://user-images.githubusercontent.com/51951532/216895589-7dba7190-e801-4176-906a-e1085a0f1d4d.png)

Before 2.15.3 there were codes that detects current orientation on camera launch but after this commit, it seems that react-native-vision-camera assumes the device is portrait on camera launch, which caused our issue.

I'll fiddle around a little more and return with what happened later.

**Comment by JackHull** [comment]
Can confirm this issue with newest version.

**Comment by mrousavy** [comment]
Can you record a video (with another phone) and post that here?

**Comment by zzz08900** [comment]
> Can you record a video (with another phone) and post that here?

Sorry I got my hands really busy these days - almost overwhelmed by customer demands. Will do this weekend.

**Comment by zzz08900** [comment]
https://user-images.githubusercontent.com/51951532/220416909-70097676-6578-433c-bd5c-2b216a0de46d.mp4


@mrousavy here's the video. You can see from the texture of the background that the content of `<Camera />` view is moving perpendicular to the movement direction of the iPad.

The App was specified to support "landscape left" and "landscape right" only in XCode and the iPad was not orientation locked, although toggling orientation lock on iPad has no effect of the outcome here.

Side note: vision camera 2.15.2 works all fine in very same configuration.

Edit: Noticed the video failed to upload, was forced to run it through iMovie/resized to a lower resolution.

**Comment by zzz08900** [comment]
Just tried adding back `private var windowInterfaceOrientation` which was removed in commit `bf71901` and nothing changed.

Do I need a full `delete everything and rebuild` after modifying files from react-native-vision-camera to make changes take effect? I got a feeling that Apple-made stuff likes to store all kinds of strange cache everywhere.

**Comment by UmerWarraich95** [comment]
Hey @zzz08900   can you look into this issue as well and guide me to solve it?
https://github.com/mrousavy/react-native-vision-camera/issues/1445

I am facing this issue on IOS camera rotation but on Android it is working fine.
If the older version of this repository can help, please let me know

**Comment by zonist** [comment]
I'm also having this issue on iphone. Android works very well.
On iphone, there is no orientation lock. When I rotate the phone to landscape, the camera view is rotated extra 90 degrees.

**Comment by UmerWarraich95** [comment]
@sekcan  I have resolved this IOS rotation issue by downgrading the react-native-vision-camera to "2.14.1"
After downgrading, my IOS build still works and I have successfully uploaded it to testflight.

**Comment by sekhuat** [comment]
I also facing the same issue, will try to downgrade to 2.14.1 if no other solution :(

**Comment by fdalmard** [comment]
I had the same problem when using version 2.15.4.
Downgrading to version 2.15.2 fixed the issue 😉

**Comment by lucksp** [comment]
> I had the same problem when using version 2.15.4. Downgrading to version 2.15.2 fixed the issue 😉


I was on version `^2.15.4` and noticing the `orientation={'portrait'}` prop doesn't seem to work when on landscape mode, in terms of how the picture is captured.  
I downgraded to 2.15.2 & 2.15.1 but had no success with the orientation changing.  MAYBE i am doing something wrong?

```
<ReanimatedCamera
                  // Do not add enableZoomGesture since using the Animated lib
                  // enableZoomGesture
                  style={{ ...styles.cameraPreview }}
                  ref={cameraRef}
                  photo
                  device={device}
                  isActive={isActive}
                  format={formats?.[0]}
                  animatedProps={animatedProps}
                  orientation={'portrait'}
                />
```

Still results as landscape when viewing final image on web too
<img width="200" alt="image" src="https://user-images.githubusercontent.com/15162169/233801116-acf9c616-d661-4558-8240-899fefc51ad5.png">

**Comment by fdalmard** [comment]
> > I had the same problem when using version 2.15.4. Downgrading to version 2.15.2 fixed the issue 😉
> 
> I was on version `^2.15.4` and noticing the `orientation={'portrait'}` prop doesn't seem to work when on landscape mode, in terms of how the picture is captured. I downgraded to 2.15.2 & 2.15.1 but had no success with the orientation changing. MAYBE i am doing something wrong?
> 
> ```
> <ReanimatedCamera
>                   // Do not add enableZoomGesture since using the Animated lib
>                   // enableZoomGesture
>                   style={{ ...styles.cameraPreview }}
>                   ref={cameraRef}
>                   photo
>                   device={device}
>                   isActive={isActive}
>                   format={formats?.[0]}
>                   animatedProps={animatedProps}
>                   orientation={'portrait'}
>                 />
> ```
> 
> Still results as landscape when viewing final image on web too <img alt="image" width="200" src="https://user-images.githubusercontent.com/15162169/233801116-acf9c616-d661-4558-8240-899fefc51ad5.png">

The `orientation` prop didn't work for me too. Here what worked for me:
- I removed the `orientation` prop
- I wrapped the Camera with a Modal with `supportedOrientations={['portrait', 'landscape']}`
- I used https://github.com/wonday/react-native-orientation-locker to force the screen orientation to either portrait (`lockToPortrait` API) or landscape (`lockToLandscapeLeft` API)

After that, the camera orientation was working as expected

```js
<Modal
  onRequestClose={() => {}}
  visible
  supportedOrientations={['portrait', 'landscape']}
>
  <Camera
    ref={cameraRef}
    style={{ flex: 1 }}
    device={device}
    isActive
    photo
    enableZoomGesture
  />
</Modal>

```

**Comment by lucksp** [comment]
thank you for suggestion @fdalmard 
The thing is, we still want to allow landscape mode (especially for video with wide angle) and not lock to an orientation, but we need the image to be saved in portrait orientation, so that viewing the result on device or on web is viewable in a portrait mode.

And I still need to test this orientation-locker package to verify.

**Comment by umbertoghio** [comment]
Hi, I have the same issue on iPhone (android working fine), I've downgraded to version 2.15.2 and the issue is solved!

Basically the rotation is 90deg wrong when landscape, it's fine when portrait, I'm unlocking screen orientation in order to let user chose portrait or landscape videos.

I tried to play with the "orientation" prop, from the swift code it seems that if present it should ignore device orientation but it made no difference, even when setting the prop dynamically on rotation event (from JS).

Hope it can help

**Comment by lucksp** [comment]
~i changed to `2.15.2` and the `orientation` is still stuck at `6` when taking pictures in both Vertical & Horizontal orientation.~

~Trying 2.14.x once again~

edit: 2.15.4 was still being installed...let me try again

update:  installed successfully, but the same outcome, where the image is not able to rotate to correctly be displayed in portrait view after it's been captured.

**Comment by lucksp** [comment]
> * I removed the `orientation` prop
> * I wrapped the Camera with a Modal with `supportedOrientations={['portrait', 'landscape']}`
> * I used https://github.com/wonday/react-native-orientation-locker to force the screen orientation to either portrait (`lockToPortrait` API) or landscape (`lockToLandscapeLeft` API)
> 
> After that, the camera orientation was working as expected
> 
> ```js
> <Modal
>   onRequestClose={() => {}}
>   visible
>   supportedOrientations={['portrait', 'landscape']}
> >
>   <Camera
>     ref={cameraRef}
>     style={{ flex: 1 }}
>     device={device}
>     isActive
>     photo
>     enableZoomGesture
>   />
> </Modal>
> ```

Thank you @fdalmard for the suggestion.  With quite a bit of custom logic, this now works when capturing an image:


```
const capturePhoto = async () => {
    if (cameraRef?.current && !snapping) {
      setSnapping(true);
      const photo = await cameraRef.current.takePhoto({
        flash: flashMode,
      });

      let rotation = 0;

      if (orientation === OrientationType['LANDSCAPE-LEFT']) {
        rotation = -90;
      } else if (orientation === OrientationType['LANDSCAPE-RIGHT']) {
        rotation = 90;
      } else if (orientation === OrientationType['PORTRAIT-UPSIDEDOWN']) {
        rotation = 180;
      }

     // now use `expo-image-manipulator`
      const photoWithOrientation = await manipulateImage({
        imageUri: photo.path,
        actions: [{ rotate: rotation }],
        saveOptions: {
          compress: 1,
        },
      });

      handleSaveAsset([
        {
          path: photoWithOrientation.uri,
          width: photoWithOrientation.width,
          height: photoWithOrientation.height,
          isRawPhoto: false,
        },
      ]);
    }
  };
```

**Comment by myselfuser1** [comment]
Try this https://www.youtube.com/playlist?list=PLQhQEGkwKZUrempLnmxjt7ZCZJu1W3p2i

**Comment by valn1** [comment]
did anyone find a solution to this?
the modal thing didn't really work for me and i don't think we should need to use that
the ios's camera should work the same as android's

**Comment by valn1** [comment]
the prop `orientation` does nothing, i wrote a setInterval to cycle through all four possibilities and the preview just blinks for a moment and logs `VisionCamera.didSetProps(_:): Updating 1 prop(s)...`
note that my app has every orientation active in the deployment info config section but for this screen i use react-native-orientation-locker to `Orientation.lockToLandscape();`
right now i am still testing stuff and trying to fix it, will bring back more information as i go
edit: `orientation` does nothing regardless of the usage of react-native-orientation-locker

**Comment by valn1** [comment]
i could not fix it the way i wanted to, but made a quick workaround.

just add the patch to your project and add the prop `orientation={'portraitUpsideDown'}` to your camera element
[react-native-vision-camera+2.15.6.patch](https://github.com/mrousavy/react-native-vision-camera/files/12468775/react-native-vision-camera%2B2.15.6.patch)

ps. i know it is not pretty, but i have deadlines

**Comment by mrousavy** [comment]
Closing as this is a stale issue - this might have been fixed with the full rewrite in VisionCamera V3 (🥳) - if not, please create a new issue.

If your issue has been fixed, consider [sponsoring me on GitHub](https://github.com/sponsors/mrousavy) to say thanks 💖

**Comment by FabianTrain** [comment]
> Closing as this is a stale issue - this might have been fixed with the full rewrite in VisionCamera V3 (🥳) - if not, please create a new issue.
> 
> If your issue has been fixed, consider [sponsoring me on GitHub](https://github.com/sponsors/mrousavy) to say thanks 💖

I recently upgraded to V3 and this issue came up. V2.15 was working for me.

**Comment by mrousavy** [comment]
Hey - I'm tracking Orientation in this feature request/issue now: https://github.com/mrousavy/react-native-vision-camera/issues/1891 ✨ 

Make sure to upvote or sponsor to support this feature, and leave a comment if you have any additional thoughts/ideas.

**Comment by umbertoghio** [comment]
> This bug is not present in version 2.14.x, I highly suspect this piece of code is what causing this issue. ![Screenshot from 2023-02-06 14-04-02](https://user-images.githubusercontent.com/51951532/216895589-7dba7190-e801-4176-906a-e1085a0f1d4d.png)
> 
> Before 2.15.3 there were codes that detects current orientation on camera launch but after this commit, it seems that react-native-vision-camera assumes the device is portrait on camera launch, which caused our issue.
> 
> I'll fiddle around a little more and return with what happened later.

Thanks to your comment I managed to resolve the orientation issue on 2.x version by forking the repository and replace iOS/CameraView+Orientation.switft with 

```
//
//  CameraView+Orientation.swift
//  VisionCamera
//
//  Created by Marc Rousavy on 04.01.22.
//  Copyright © 2022 mrousavy. All rights reserved.
//

import Foundation
import UIKit

extension CameraView {
  /// Returns the current _interface_ orientation of the main window
  private var windowInterfaceOrientation: UIInterfaceOrientation {
    if #available(iOS 13.0, *) {
      return UIApplication.shared.windows.first?.windowScene?.interfaceOrientation ?? .unknown
    } else {
      return UIApplication.shared.statusBarOrientation
    }
  }

  /// Orientation of the input connection (preview)
  private var inputOrientation: UIInterfaceOrientation {
    return windowInterfaceOrientation
  }

  // Orientation of the output connections (photo, video, frame processor)
  private var outputOrientation: UIInterfaceOrientation {
    if let userOrientation = orientation as String?,
       let parsedOrientation = try? UIInterfaceOrientation(withString: userOrientation) {
      // user is overriding output orientation
      return parsedOrientation
    } else {
      // use same as input orientation
      return inputOrientation
    }
  }

  internal func updateOrientation() {
    // Updates the Orientation for all rotable connections (outputs) as well as for the preview layer
    DispatchQueue.main.async {
      // `windowInterfaceOrientation` and `videoPreviewLayer` should only be accessed from UI thread
      let isMirrored = self.videoDeviceInput?.device.position == .front

      self.videoPreviewLayer.connection?.setInterfaceOrientation(self.inputOrientation)

      let connectionOrientation = self.outputOrientation
      self.cameraQueue.async {
        // Run those updates on cameraQueue since they can be blocking.
        self.captureSession.outputs.forEach { output in
          output.connections.forEach { connection in
            if connection.isVideoMirroringSupported {
              connection.automaticallyAdjustsVideoMirroring = false
              connection.isVideoMirrored = isMirrored
            }
            connection.setInterfaceOrientation(connectionOrientation)
          }
        }
      }
    }
  }
}

```

**Comment by andrespompiglio** [comment]
> > This bug is not present in version 2.14.x, I highly suspect this piece of code is what causing this issue. ![Screenshot from 2023-02-06 14-04-02](https://user-images.githubusercontent.com/51951532/216895589-7dba7190-e801-4176-906a-e1085a0f1d4d.png)
> > Before 2.15.3 there were codes that detects current orientation on camera launch but after this commit, it seems that react-native-vision-camera assumes the device is portrait on camera launch, which caused our issue.
> > I'll fiddle around a little more and return with what happened later.
> 
> Thanks to your comment I managed to resolve the orientation issue on 2.x version by forking the repository and replace iOS/CameraView+Orientation.switft with
> 
> ```
> //
> //  CameraView+Orientation.swift
> //  VisionCamera
> //
> //  Created by Marc Rousavy on 04.01.22.
> //  Copyright © 2022 mrousavy. All rights reserved.
> //
> 
> import Foundation
> import UIKit
> 
> extension CameraView {
>   /// Returns the current _interface_ orientation of the main window
>   private var windowInterfaceOrientation: UIInterfaceOrientation {
>     if #available(iOS 13.0, *) {
>       return UIApplication.shared.windows.first?.windowScene?.interfaceOrientation ?? .unknown
>     } else {
>       return UIApplication.shared.statusBarOrientation
>     }
>   }
> 
>   /// Orientation of the input connection (preview)
>   private var inputOrientation: UIInterfaceOrientation {
>     return windowInterfaceOrientation
>   }
> 
>   // Orientation of the output connections (photo, video, frame processor)
>   private var outputOrientation: UIInterfaceOrientation {
>     if let userOrientation = orientation as String?,
>        let parsedOrientation = try? UIInterfaceOrientation(withString: userOrientation) {
>       // user is overriding output orientation
>       return parsedOrientation
>     } else {
>       // use same as input orientation
>       return inputOrientation
>     }
>   }
> 
>   internal func updateOrientation() {
>     // Updates the Orientation for all rotable connections (outputs) as well as for the preview layer
>     DispatchQueue.main.async {
>       // `windowInterfaceOrientation` and `videoPreviewLayer` should only be accessed from UI thread
>       let isMirrored = self.videoDeviceInput?.device.position == .front
> 
>       self.videoPreviewLayer.connection?.setInterfaceOrientation(self.inputOrientation)
> 
>       let connectionOrientation = self.outputOrientation
>       self.cameraQueue.async {
>         // Run those updates on cameraQueue since they can be blocking.
>         self.captureSession.outputs.forEach { output in
>           output.connections.forEach { connection in
>             if connection.isVideoMirroringSupported {
>               connection.automaticallyAdjustsVideoMirroring = false
>               connection.isVideoMirrored = isMirrored
>             }
>             connection.setInterfaceOrientation(connectionOrientation)
>           }
>         }
>       }
>     }
>   }
> }
> ```

Thanks!, we've applied above change to version 2.16.3 and it worked ok!

**Comment by garmoshka-mo** [comment]
there is still issue on iPads - if you will try to rotate device while camera is open - it will turn image improperly.

You need to add this to `outputOrientation` to fix camera on iPads:

```
  private var outputOrientation: UIInterfaceOrientation {
    if (UIDevice.current.userInterfaceIdiom == .pad) {
        return inputOrientation
    }
```

Also on some iPads (e.g. iPad Air5 iOS17) UI orientation value is lagging from device orientation, thus orientation change listener actually should be on ViewController's.viewWillTransition, but this requires better knowledge of swift (to embed controller to controller's chain on mount).
As hotfix I renamed to `_updateOrientation` and did:

```
  internal func updateOrientation() {
      _updateOrientation()
      // hotfix: for some reason on iPad Air5 iOS17 UI orientation value is lagging from device orientation changing event. (e.g. because of larger screen, iOS17 or else...)
      DispatchQueue.main.asyncAfter(deadline: .now() + 0.3) {
          self._updateOrientation()
          DispatchQueue.main.asyncAfter(deadline: .now() + 0.7) {
              self._updateOrientation()
          }
      }
  }
```
